### PR TITLE
Add TypeScript, catch bugs

### PR DESCRIPTION
**Pull request by bcherny**
This PR:

- Adds TypeScript, TSLint, and TSJest.
- Switches the JS build pipeline from Babel to TSC (alternatively, we can use both, but since the UI is so simple, it's simpler to just use TSC without Babel). TSC now compiles both .ts and .js/.jsx files.
- Ports Links.jsx to TypeScript, catching a few bugs along the way.

-------

**Tests pass, but I'd love help testing this end to end. It's likely that this doesn't work at runtime yet, and will require a little more tweaking.**

### CODE REVIEW

File: gdbgui/src/js/Links.tsx [hunk 1]
```diff
@@ -1,14 +1,22 @@
...
+  fullname?: string
+  line: string
+  num_lines?: number
+}
+
+export class FileLink extends React.Component<Props> {
   render() {
     let line = parseInt(this.props.line);
-    let onclick = null,
+    let onclick = () => {},
```
Comment (bcherny) [inline gdbgui/src/js/Links.tsx]: Bug #2: React event handlers shouldn't be null.

File: gdbgui/src/js/Links.tsx [hunk 2]
```diff
@@ -1,14 +1,22 @@
...
       cls = "";
     if (!this.props.file || !line) {
-      line = "";
+      line = 0;
```
Comment (bcherny) [inline gdbgui/src/js/Links.tsx]: Bug #3: `line` is a number, not a string.

File: gdbgui/src/js/Links.tsx [hunk 3]
```diff
@@ -1,14 +1,22 @@
 import Actions from "./Actions.js";
 import React from "react";
 import CopyToClipboard from "./CopyToClipboard.jsx";
+import MemoryLink from "./MemoryLink.jsx"
```
Comment (bcherny) [inline gdbgui/src/js/Links.tsx]: Bug #1: It looks like we forgot to import `MemoryLink` before?

**Comment by cs01** [comment]
Thank you for this! It's amazing. 

A couple things:

* the build now takes ~25 sec /  3 sec for build/rebuild on my computer (it used to take ~9 sec / < 1 sec). Would bringing babel back in make this faster?
* I see this in the build now
> No valid rules have been specified for JavaScript files
> No valid rules have been specified for TypeScript files

Is this expected?

**Comment by bcherny** [comment]
Hey @cs01! I just came back from vacation.

> the build now takes ~25 sec / 3 sec for build/rebuild on my computer (it used to take ~9 sec / < 1 sec). Would bringing babel back in make this faster?

That's definitely longer than expected. Will take a look in the next 1-2 weeks.

> I see this in the build now

Looks like I misconfigured something in Webpack. Will take a look at that one too, probably a 1-liner.

**Comment by bcherny** [comment]
@cs01 Updated!

**Comment by cs01** [comment]
Thanks! Will take a look when I get the time.

**Comment by cs01** [comment]
I tried it out. It builds with no errors, and finds type errors while it's watching with `yarn start`. Awesome!

The build time went from 5814ms to ~30 for intial build, but incremental builds are only ~1 second, so I can live with that.

It looks like the source maps were removed, was there a reason for that? I would prefer to keep them if possible.

File: webpack.config.js [hunk 4]
```diff
@@ -1,31 +1,42 @@
 const path = require('path');
+const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin');
 
 module.exports = {
   entry: './gdbgui/src/js/gdbgui.jsx',
-  devtool: 'source-map',
```
Comment (cs01) [inline webpack.config.js]: does this have to be removed?

**Comment by cs01** [comment]
Looking around some more, and it seems like the watcher is a little flaky. have you encountered this? I'm not sure if it's something wrong with my setup or not. 

If I run `yarn start`, after ~30 seconds the build finishes and I see this

```
> yarn start
yarn run v1.15.2
$ NODE_ENV=development webpack --mode development --watch --config webpack.config.js
Starting type checking and linting service...
Using 1 worker with 2048MB memory limit

Webpack is watching the files…

No type errors found
No lint errors found
Version: typescript 3.3.3333, tslint 5.13.1
Time: 12553ms
Hash: 985759485b38bf6b9883
Version: webpack 4.12.1
Time: 28367ms
Built at: 03/21/2019 5:42:34 PM
       Asset      Size  Chunks             Chunk Names
    build.js  1.11 MiB    main  [emitted]  main
build.js.map  1.24 MiB    main  [emitted]  main
[./gdbgui/src/js/Actions.js] 11.2 KiB {main} [built]
[./gdbgui/src/js/FileOps.jsx] 25.4 KiB {main} [built]
[./gdbgui/src/js/FoldersView.jsx] 12.1 KiB {main} [built]
[./gdbgui/src/js/GdbApi.jsx] 15.3 KiB {main} [built]
[./gdbgui/src/js/GdbConsoleContainer.jsx] 6.84 KiB {main} [built]
[./gdbgui/src/js/GdbguiModal.jsx] 2.08 KiB {main} [built]
[./gdbgui/src/js/GlobalEvents.js] 2.18 KiB {main} [built]
[./gdbgui/src/js/HoverVar.jsx] 4.46 KiB {main} [built]
[./gdbgui/src/js/RightSidebar.jsx] 8.27 KiB {main} [built]
[./gdbgui/src/js/Settings.jsx] 5.53 KiB {main} [built]
[./gdbgui/src/js/ToolTip.jsx] 2.64 KiB {main} [built]
[./gdbgui/src/js/ToolTipTourguide.jsx] 5.49 KiB {main} [built]
[./gdbgui/src/js/TopBar.jsx] 16.1 KiB {main} [built]
[./gdbgui/src/js/constants.js] 2.03 KiB {main} [built]
[./gdbgui/src/js/gdbgui.jsx] 6.74 KiB {main} [built]
    + 50 hidden modules

```

but when I change a file, nothing happens. I saw this earlier, but killed the `yarn start` process, then restarted and it worked so I thought it was my imagination. But it is happening again now.

**Comment by cs01** [comment]
Thanks again @bcherny!

**Pull request merged**